README.md
# simple-manga-downloader
Simple manga console downloader written in python. Made for Linux but should work on any platform.

Currently supports:

- [mangadex](https://mangadex.org)
- [mangatown](https://www.mangatown.com)
- [mangakakalot](https://mangakakalot.com/page)
- [manganato](https://manganato.com/)

*Note:* mangadex is the main supported site, the others *might* be maintained but they are not my top priority.

Allows you to download manga in 5 ways:

- all chapters of a series
- a range of chapters (from 5 to 15)
- a selection of chapters (5, 7, 10, 20)
- only the newest chapter
- check for new chapters for the tracked manga

Additional features of the downloader:

- It will check if a chapter on mangadex.org has multiple uploads by different groups and ask which one to download
- It handles the MangaPlus chapters on mangadex
- Allows to use data saver images from Mangadex
- It can download the cover(s) for the manga (off by default)
- You can specify the language (mangadex only)
- You can specify which chapters to exclude from the download
- You can change the directory where the manga is saved
- You can add ongoing manga to the tracked list for a easy way to check for new chapters
- Config is saved as a .json for readability and easy modification
- The downloader has a config "mode" that allows the modification of the config file without having to edit the .json manually
- It can check for new available versions
- It will remove (or replace) characters from titles that could cause problems, by default removes `/ \ | ? > < . : ? * "`


# Installation

## Requirements
- BeautifulSoup 4
- requests

## Releases form PyPI

Installing using pip will handle the requirements automatically. I would also recommend using [pipx](https://github.com/pipxproject/pipx) instead of just pip.

```
pip install simple-manga-downloader
```

## Using the git repo

### Installing using pip

To install from the master branch

```
pip install git+https://github.com/Kanjirito/simple-manga-downloader@master
```

### Using the git repo without installing

Copy the repo, install the requirements and use the entry script in the main repo directory
```
cd path/to/repo
chmod +x SMD.py
pip install -r requirements.txt
./SMD.py [mode] [arguments]
```


# USAGE

## General info

For all commands you can either use `SMD` or `smd`.

The default manga download directory is `~/Manga` this can be changed in the config file or in the command. The config directory is `$XDG_CONFIG_HOME/SMD` (`~/.config/SMD` if variable not set), you can specify a different config to be used like this:

```
SMD -c "path/to/config" mode arguments
```

Examples:
```
SMD -c "~/Downloader/config.json" down link_to_manga -l
SMD -c "~/Downloader/config.json" update
```
To create the config file the downloader needs to successfully finish. If you want to have the config created before you use the downloader you can do:
```
SMD conf
```

## Details
For a more detailed explanation of the modes and their available arguments read the [USAGE.md](USAGE.md) file in the git repo.

# Changelog

Changelog can be found in [CHANGELOG.md](CHANGELOG.md) in the git repo.


CHANGELOG.md
# Release History

## v1.16.1(2022-06-11)

### Fixes
- Adds a replacement rule for `"`

## v1.16.0(2021-12-30)

### New
- `smd` entry point for convenience. `SMD` will still work

### Fixes
- Chapter hash and page names are now at `/at-home`

## v1.15.5(2021-10-09)

### Fixes
- Mangadex works again (doesn't fetch groups separately any more)
- Changing a replacement rule with no string given now sets it to " " which is in line with the default rules

## v1.15.4(2021-07-30)

### Changes
- Add support for Manganelo links for Manganato (this might change at any point)
- The default replacement rules change to space instead of nothing

### Internal
- Better flake8 rule selection

## v1.15.3(2021-07-26)

### Changes
- Manganelo changed name to Manganato

## v1.15.2(2021-06-07)

### Fixes
- Crash caused by outdated urllib3

## v1.15.1(2021-05-26)

### Fixes
- Crash when getting some mangadex covers

## v1.15.0(2021-05-25)

### New
- Mangadex covers (they also now respect the data saver option)

### Fixes
- Rate limiting not actually working as it should
- Better message on rate limit error
- Mangadex results are now also sorted by volume (should help a bit with anthologies)

### Code-quality
- `cover_url` renamed to `covers`

## v1.14.0(2021-05-22)

### Fixes
- mangadex language selection (API changed the query parameter)
- Typo that would cause crash if downloader was ran directly

### Changes
- Very slight help message changes

### Internal
- Code moved to follow [black](https://black.readthedocs.io/en/stable/) formatting
- Added flake8 with plugins to catch mistakes
- Started using isort to automatically sort imports
- Added `requirements-dev.txt`

## v1.13.0(2021-05-21)

### Changes
- Migrated to the new Mangadex API
- Slight output changes
- Downloader will now exit with status code 1 if config has a invalid language code

### Removed
- MD@Home setting

### New
- Data saving setting `conf --data-saver` that will download the lower quality images (off by default)
- `--data-saver {yes(y),no(n)}` flag for `down` and `update` that overwrites the data saver setting in the config file
- `down --language` and `update --language` arguments that allow to overwrite the config language setting

### Fixes
- Mangadex no longer says that there's been an error when no chapters found matching the criteria but just prints that none were found
- Crash on python 3.6 because of unsupported `required` keyword argument in `add_subparsers`

## v1.12.0(2021-01-20)

### New
- MD@Home setting
- basic config validation

### Fixes
- mangakakalot crash
- crash when the link points to a non existing manga

### Changes
- Mangadex module now also works with URLs like `/manga/ID`
- downloader now checks the chapters in ascending order (won't effect much)
- downloader now asks for a new chapter number if it's already taken (fixes #21, might cause unexpected behaviour with `-c/--check` or `-i/--ignore-input`)

## v1.11.2(2021-01-11)

### Fixes
- version check now properly compares to the newest version

## v1.11.1(2021-01-11)

### Changes
- removed dot from cover filename for mangadex

## v1.11.0(2021-01-11)

### Changes
- migrated to the Mangadex V2 API
- downloader will now get all available covers (only mangadex supports this for now)

## v1.10.0(2020-10-22)

### New
- if manga is tracked the tracked title will be used instead of the fetched one
- title is now determined in order of: custom name > tracked title > fetched title
- you can now use a tracked title as a download target in down mode
- new `conf --change-name` argument for changing the title of tracked manga
- when renaming tracked manga downloader will ask if it should rename old files
- new `conf -n/--name` argument to allow adding manga to tracked with a custom name (should not be used when adding multiple manga)
- character replacement list for titles, by default replaces `/ \ | ? > < . : ? *` with nothing (just removes them)
- new conf arguments for the replacement rules (`--rule-reset/--rule-print/--rule-add/--rule-remove`)

### Changes
- help message changed to fit new functionality
- changed all long argument to use "-" instead of "\_"
- config will now look for the environment variable `XDG_CONFIG_HOME` before defaulting to `~/.config` 
- white space removed from beginning and end of titles (only really matters for custom titles)

### Fixes
- downloader no longer catches all exceptions (oops)

## v1.9.3(2020-09-07)

### Fixes
- Manga Plus releases not being handled correctly

## v1.9.2(2020-07-07)

### Removed
- mangaseeonline support, site now requires JS to work

### Changes
- internal stuff

## v1.9.1(2020-06-26)

### Fixes
- manganelo and mangakakalot image downloads no longer fail because of cloudflare

## v1.9.0(2020-06-21)

### Fixes
- fixed crash if manga had "/" in title, now is replaced with "╱"

### Changes
- `-i/--ignore_input` and `-c/--check` are now mutually exclusive
- removed titles from mangadex links
- `conf -r/--remove-tracked` uses the site regex to check if link can be shortened (for now only mangadex benefits from this)
- preserved order of input for `down`, `conf -a/--add-tracked`
- more verbose output for `conf -r/--remove-tracked` if argument is not tracked
- `-a/--add-tracked` and `-r/--remove-tracked` are now mutually exclusive giving an error instead of just doing one
- `-s/--save-directory` now prints a confirmation that the directory got changed
- config is now always saved after KeyboardInterrupt
- failing to load config will result in exit code 1

### Internal
- all site modules now inherit from BaseManga
- all site modules now create a single `requests.Session()` used by all instances of it
- argument parsing moved to it's own module
- argument parser and config parser moved out of modules directory
- better checking if manga in tracked
- using regex to find module for link allowing for nicer module importing
- functions for setting `BaseManga` attributes added for clarity

## v1.8.0(2020-06-04)

### New

simple_manga_downloader/arg_parser.py
import argparse


def parse_arguments():
    """Parses the arguments
    Returns the arguments object
    """

    class SetAction(argparse.Action):
        """Action for argparse that creates a set"""

        def __call__(self, parser, namespace, values, option_string=None):
            setattr(namespace, self.dest, set(values))

    class NoDupesOrderedListAction(argparse.Action):
        """Action for argparse that creates a list with no dupes and preserved order"""

        def __call__(self, parser, namespace, values, option_string=None):
            setattr(namespace, self.dest, list(dict.fromkeys(values)))

    parser = argparse.ArgumentParser(
        description=(
            "SMD is a command line manga downloader. For more information read the README file in the GitHub repo.\n"
            "https://github.com/Kanjirito/simple-manga-downloader/blob/master/README.md"
        ),
        formatter_class=argparse.RawDescriptionHelpFormatter,
    )

    parser.add_argument(
        "-c",
        "--custom",
        dest="custom_cfg",
        metavar="PATH/TO/CONFIG",
        help="Sets a custom config to use",
        default=None,
    )

    # Sub-parsers for the different modes
    subparsers = parser.add_subparsers(dest="subparser_name", metavar="mode")
    # Has to be this way for python 3.6 support
    subparsers.required = True

    parser_conf = subparsers.add_parser(
        "conf",
        help="Downloader will be in config edit mode",
        description="Changes the settings",
    )
    parser_down = subparsers.add_parser(
        "down",
        help="Downloader will be in download mode",
        description="Downloads the given manga. Supports multiple links at once.",
    )
    parser_update = subparsers.add_parser(
        "update",
        help="Downloader will be in update mode",
        description="Download new chapters from tracked list",
    )
    parser_version = subparsers.add_parser(
        "version",
        help="Downloader will be in version mode",
        description="Prints the downloader version",
    )

    # Parser for download mode
    parser_down.add_argument(
        "input",
        nargs="+",
        metavar="target [another_target]",
        action=NoDupesOrderedListAction,
        help="Manga URL, tracked index or tracked title to download",
    )
    parser_down.add_argument(
        "-d",
        "--directory",
        dest="custom_down_dire",
        metavar="PATH/TO/DIRECTORY",
        default=None,
        help="Custom path for manga download",
    )
    parser_down.add_argument(
        "-e",
        "--exclude",
        help="Chapters to exclude",
        metavar="NUMBER",
        nargs="+",
        type=float,
        action=SetAction,
        default=set(),
    )
    parser_down.add_argument(
        "-n",
        "--name",
        default=None,
        metavar="CUSTOM NAME",
        nargs="+",
        help="Download the manga with a custom name. Not recommended to use with multiple downloads at once.",
    )
    parser_down.add_argument(
        "--data-saver",
        help="Toggle the Mangadex data saver option. Takes priority over config file.",
        choices=["y", "yes", "n", "no"],
        metavar=r"{yes(y),no(n)}",
        dest="data_saver",
    )
    parser_down.add_argument(
        "--language",
        help="Overwrite the Mangadex language setting",
        dest="langauge_code",
    )

    input_group_down = parser_down.add_mutually_exclusive_group()
    input_group_down.add_argument(
        "-c",
        "--check",
        help="Only check for new chapters without downloading and asking for any input",
        action="store_true",
        dest="check_only",
    )
    input_group_down.add_argument(
        "-i",
        "--ignore-input",
        help="Downloads without asking for any input in case of group conflict the alphabetically first group will be chosen)",
        action="store_true",
        dest="ignore_input",
    )

    selection_group = parser_down.add_mutually_exclusive_group()
    selection_group.add_argument(
        "-r",
        "--range",
        help="Specifies the range of chapters to download, both ends are inclusive.",
        metavar=("START", "END"),
        nargs=2,
        type=float,
    )
    selection_group.add_argument(
        "-s",
        "--selection",
        help='Specifies which chapters to download. Accepts multiple chapters "2 10 25"',
        metavar="NUMBER",
        nargs="+",
        action=SetAction,
        type=float,
    )
    selection_group.add_argument(
        "-l", "--latest", help="Download only the latest chapter", action="store_true"
    )

    # Parser for config mode
    tracked_edit_group = parser_conf.add_mutually_exclusive_group()
    tracked_edit_group.add_argument(
        "-a",
        "--add-tracked",
        help="Adds manga to the tracked list",
        dest="add_to_tracked",
        metavar="MANGA_URL",
        nargs="+",
        action=NoDupesOrderedListAction,
    )
    tracked_edit_group.add_argument(
        "-r",
        "--remove-tracked",
        help="Removes manga from tracked. Supports deletion by url, title or tracked index",
        dest="remove_from_tracked",
        metavar="{MANGA_URL,MANGA_TITLE,NUMBER}",
        nargs="+",
        action=SetAction,
    )

    parser_conf.add_argument(
        "-n",
        "--name",
        help="When adding manga give it a custom name (do not use when adding multiple at once)",
        dest="name",
        default=None,
        nargs="+",
        metavar="CUSTOM NAME",
    )
    parser_conf.add_argument(
        "-t", "--clear-tracked", help="Clears the tracked list", action="store_true"
    )
    parser_conf.add_argument(
        "-s",
        "--save-directory",
        help="Changes the manga download directory",
        metavar="PATH/TO/DIRECTORY",
        dest="manga_down_directory",
    )
    parser_conf.add_argument(
        "-d", "--default", help="Resets the config to defaults", action="store_true"
    )
    parser_conf.add_argument(
        "-l",
        "--list-tracked",
        help="Lists all of the tracked manga",
        action="store_true",
    )
    parser_conf.add_argument(
        "-m",
        "--modify-position",
        help="Changes the position of tracked manga",
        action="store_true",
        dest="modify_tracked_position",
    )
    parser_conf.add_argument(
        "-v",
        "--verbose",
        help="Used with -l or -m to also print links",
        action="store_true",
        dest="verbose",
    )
    parser_conf.add_argument(
        "--change-name",
        help="Change the name of a tracked manga",
        action="store_true",
        dest="change_name",
    )
    parser_conf.add_argument(
        "-p",
        "--print-conf",
        help="Print all config settings",
        action="store_true",
        dest="print_config",
    )
    parser_conf.add_argument(
        "-c",
        "--covers",
        help="Toggles the cover download setting",
        action="store_true",
        dest="toggle_covers",
    )
    parser_conf.add_argument(
        "--change-lang",
        help="Changes the mangadex language code",
        metavar="LANGUAGE CODE",
    )
    parser_conf.add_argument(
        "--list-lang",
        help="Lists all of the mangadex language codes",
        action="store_true",
    )
    parser_conf.add_argument(
        "--timeout",
        help="Change the download timeout",
        metavar="SECONDS",
        type=int,
        dest="timeout",
    )
    parser_conf.add_argument(
        "--rule-reset",
        help="Resets the replacement rules to the default",
        action="store_true",
        dest="replacement_reset",
    )
    parser_conf.add_argument(
        "--rule-print",
        help="Prints the current replacement rules",
        action="store_true",
        dest="rule_print",
    )
    parser_conf.add_argument(
        "--data-saver",
        help="Toggles the date saving setting",
        action="store_true",
        dest="data_saver",
    )

    replacement_rules_group = parser_conf.add_mutually_exclusive_group()
    replacement_rules_group.add_argument(
        "--rule-add",
        help="Adds a new replacement rule for a character. If no replacement given it will default to removing it.",
        nargs="+",
        metavar=("CHARACTER", "REPLACEMENT"),
        dest="rule_add",
    )
    replacement_rules_group.add_argument(
        "--rule-remove",
        help="Removes a replacement rule for a character",
        metavar="CHARACTER",
        dest="rule_remove",
    )

    # Update options
    input_group_update = parser_update.add_mutually_exclusive_group()
    input_group_update.add_argument(
        "-c",
        "--check",
        help="Only check for new chapters without downloading and asking for any input",
        action="store_true",
        dest="check_only",
    )
    input_group_update.add_argument(
        "-i",
        "--ignore-input",
        help="Downloads without asking for any input (in case of group conflict the alphabetically first group will be chosen)",
        action="store_true",
        dest="ignore_input",
    )

    parser_update.add_argument(
        "-d",
        "--directory",
        dest="custom_down_dire",
        metavar="PATH/TO/DIRECTORY",
        default=None,
        help="Custom path for manga download",
    )
    parser_update.add_argument(
        "--language",
        help="Overwrite the Mangadex language setting",
        dest="langauge_code",
    )
    parser_update.add_argument(
        "--data-saver",
        help="Toggle the Mangadex data saver option. Takes priority over config file.",
        choices=["y", "yes", "n", "no"],
        metavar=r"{yes(y),no(n)}",
        dest="data_saver",
    )

    # Version options
    parser_version.add_argument(
        "-c",
        "--check",
        help="Checks if there is a new version available",
        action="store_true",
        dest="version_check",
    )

    return parser.parse_args()
